src/main/vcpf4/usbd_cdc_vcp.c

```diff
@@ -193,12 +192,17 @@ uint32_t CDC_Receive_DATA(uint8_t* recvBuf, uint32_t len)
 {
     uint32_t count = 0;
 
-    while (usbAvailable() && count < len) {
-        recvBuf[count] = usbData.buffer[usbData.bufferOutPosition];
-        usbData.bufferOutPosition = (usbData.bufferOutPosition + 1) % USB_RX_BUFSIZE;
+    while (APP_Tx_ptr_out != APP_Tx_ptr_in && count < len) {
+        recvBuf[count] = APP_Tx_Buffer[APP_Tx_ptr_out];
+        APP_Tx_ptr_out = (APP_Tx_ptr_out + 1) % USB_RX_BUFSIZE;
         count++;
         receiveLength--;
     }
+
+    if (!receiveLength) {
+        receiveLength = APP_Tx_ptr_out != APP_Tx_ptr_in;
```

Comment:
It does. Sort of. The only use of receiveLength is to check in serial VCP as to whether there's anything in the buffer. i.e. receiveLength > 0. The issue is it can become 0, when there's something in the buffer - and thus this never gets called again (a pseudo lockup). This is safety.